rclcpp/include/rclcpp/create_timer.hpp

```diff
@@ -132,12 +114,97 @@ create_wall_timer(
             "Casting timer period to nanoseconds resulted in integer overflow."};
   }
 
+  return period_ns;
+}
+
+/// Convenience method to create a wall timer with node resources.
+/**
+ *
+ * \tparam DurationRepT
+ * \tparam DurationT
+ * \tparam CallbackT
+ * \param period period to execute callback. This duration must be 0 <= period < nanoseconds::max()
+ * \param callback callback to execute via the timer period
+ * \param group callback group
+ * \param node_base node base interface
+ * \param node_timers node timer interface
+ * \return shared pointer to a wall timer
+ * \throws std::invalid_argument if either node_base or node_timers
+ * are null, or period is negative or too large
+ */
+template<typename DurationRepT, typename DurationT, typename CallbackT>
+typename rclcpp::WallTimer<CallbackT>::SharedPtr
+create_wall_timer(
+  std::chrono::duration<DurationRepT, DurationT> period,
+  CallbackT callback,
+  rclcpp::CallbackGroup::SharedPtr group,
+  node_interfaces::NodeBaseInterface * node_base,
+  node_interfaces::NodeTimersInterface * node_timers)
+{
+  if (node_base == nullptr) {
+    throw std::invalid_argument{"input node_base cannot be null"};
+  }
+
+  if (node_timers == nullptr) {
+    throw std::invalid_argument{"input node_timers cannot be null"};
+  }
+
+  const std::chrono::nanoseconds period_ns = safe_cast_to_period_in_ns(period);
+
+  // Add a new wall timer.
   auto timer = rclcpp::WallTimer<CallbackT>::make_shared(
     period_ns, std::move(callback), node_base->get_context());
   node_timers->add_timer(timer, group);
   return timer;
 }
 
+/// Convenience method to create a general timer with node resources.
+/**
+ *
+ * \tparam DurationRepT
+ * \tparam DurationT
+ * \tparam CallbackT
+ * \param period period to execute callback. This duration must be 0 <= period < nanoseconds::max()
+ * \param callback callback to execute via the timer period
+ * \param group callback group
+ * \param node_base node base interface
+ * \param node_timers node timer interface
+ * \return shared pointer to a generic timer
+ * \throws std::invalid_argument if either node_base, node_timers or node_clock
+ * are null, or period is negative or too large
+ */
+template<typename DurationRepT, typename DurationT, typename CallbackT>
+typename rclcpp::GenericTimer<CallbackT>::SharedPtr
+create_timer(
```

Comment:
I think this needs to be tested in `rclcpp/test/rclcpp/test_create_timer.cpp`?

Reply:
Looks like [simply including this header](https://github.com/ros2/rclcpp/pull/1975/commits/4594e88330dbfc3ce98e81f9f10c5d02c3737742#diff-cb1652cc3b1eff7b2611c2b78644d80d73b7e7d4185649e20a88926de473af2dR26) has triggered a CI build failure, which might be beyond my ability to solve...

Reply:
> Looks like [including this header](https://github.com/ros2/rclcpp/pull/1975/commits/4594e88330dbfc3ce98e81f9f10c5d02c3737742#diff-cb1652cc3b1eff7b2611c2b78644d80d73b7e7d4185649e20a88926de473af2dR26) has triggered a CI failure, which might be beyond my ability to solve. I can take a look but I might need @jacobperron to step in.

It looks unlikely to be because of including the header.  Probably something in here is causing an endless loop, or at least a very slow loop, so that it hits the timeout.

Reply:
Oh, that's interesting. Thanks, @clalancette.  I had completely misinterpreted the `[ERROR] [1659380009.202064912] [rclcpp]: failed to remove clock` as being the source of the issue. It seems too coincidental that the test took exactly one minute and zero seconds to complete. Perhaps my additional tests pushed it over some 60s internal threshold. For what it's worth, the tests pass locally so perhaps just bumping the timeout might fix the issue.

Reply:
It's a little hard to see, but if you search for `test_time_source` in https://build.ros2.org/job/Rpr__rclcpp__ubuntu_jammy_amd64/215/consoleFull (for example), you can see it currently takes ~18 seconds.  Going beyond 60 seconds here is a huge increase, so I'd definitely say that there is something weird going on in this PR.

Reply:
OK -- Now that I can see a full console log delta between builds pre- and post- my changes, it's likely that this PR is interacting negatively with simulated time in replay -- or its potentially surfacing a pre-existing issue. This agrees with local rosbag2 replay behavior I'm seeing, and might point to a larger issue with node timers driven by the `/clock` topic. I'll try and dig deeper to work out why.

Reply:
@clalancette do you know if [this line of code](https://github.com/ros2/rclcpp/blob/4594e88330dbfc3ce98e81f9f10c5d02c3737742/rclcpp/test/rclcpp/test_time_source.cpp#L675) is correct? To me it looks like nanoseconds is being incremented and will overflow. Should it not rather be something like this:

```c++
pub_time_ += rclcpp::Duration(0, 1000000);
clock_msg_.clock = pub_time_;
clock_pub_->publish(clock_msg_);
```

Reply:
Off-hand, I'm not sure.  That said, I think your change makes sense, so I'd give that a shot.

Reply:
All failed tests happened on [agent-3af8c62f](https://build.ros2.org/computer/agent-3af8c62f). I think this CI worker needs to be spun down, as it's causing issues.

Reply:
I commented out one test (see https://github.com/ros2/rclcpp/pull/1975#discussion_r956348057), that isn't doing what it's supposed to do.
The other tests that are flaky, I was able to reproduce the failure both for this branch and for `rolling`, so the issue is unrelated to this work.